compiler/rustc_mir_build/src/build/matches/test.rs

```diff
@@ -282,13 +282,20 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                         place,
                         ty,
                     );
-                } else if let [success, fail] = *make_target_blocks(self) {
-                    assert_eq!(value.ty(), ty);
-                    let expect = self.literal_operand(test.span, value);
-                    let val = Operand::Copy(place);
-                    self.compare(block, success, fail, source_info, BinOp::Eq, expect, val);
                 } else {
-                    bug!("`TestKind::Eq` should have two target blocks");
+                    let [success, fail] = *make_target_blocks(self) else {
+                        bug!("`TestKind::Eq` should have two target blocks");
+                    };
+                    assert_eq!(value.ty(), ty);
+                    if matches!(ty.kind(), ty::FnDef(..)) {
```

Comment:
`rustc_borrowck::type_check` only allows comparisons for a few types. Should we use the same list of allowed types here, instead of special-casing `FnDef`?

Reply:
That list seems to just be `Ty::is_scalar` except for `FnDef`?

I actually don't really even know why `FnDef` is counted as scalar in `Ty::is_scalar`. Seems a bit misleading at best.

Reply:
Though I don't totally understand what you mean here -- are you suggesting this should be an error instead?